flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/legacy/metrics/AbstractAggregatingMetricsHandler.java

```diff
@@ -0,0 +1,223 @@
+									break;
+								case DoubleAccumulator.DoubleAverage.NAME:
+									requestedAggregationsFactories.add(DoubleAccumulator.DoubleAverageFactory.get());
+									break;
+								default:
+									log.warn("Invalid aggregation specified: {}", aggregation.toLowerCase());
+							}
+						}
+					}
+					Collection<? extends MetricStore.ComponentMetricStore> stores = getStores(store, pathParams, queryParams);
+					if (stores == null){
+						return "[]";
+					}
+					return getAggregatedMetricValues(stores, requestedMetrics, requestedAggregationsFactories);
+				} catch (Exception e) {
+					throw new CompletionException(new FlinkException("Could not retrieve metrics.", e));
+				}
+			},
+			executor);
+	}
+
+	/**
+	 * Returns a JSON string containing a list of all available metrics in the given stores. Effectively this method maps
+	 * the union of all key-sets to JSON.
+	 *
+	 * @param stores metrics
+	 * @return JSON string containing a list of all available metrics
+	 */
+	private static Collection<String> getAvailableMetrics(Collection<? extends MetricStore.ComponentMetricStore> stores) {
+		Set<String> uniqueMetrics = new HashSet<>();
+		for (MetricStore.ComponentMetricStore store : stores) {
+			uniqueMetrics.addAll(store.metrics.keySet());
+		}
+		return uniqueMetrics;
+	}
+
+	private static String mapMetricListToJson(Collection<String> metrics) throws IOException {
+		StringWriter writer = new StringWriter();
+		JsonGenerator gen = JsonFactory.JACKSON_FACTORY.createGenerator(writer);
+
+		gen.writeStartArray();
+		for (String m : metrics) {
+			gen.writeStartObject();
+			gen.writeStringField("id", m);
+			gen.writeEndObject();
+		}
+		gen.writeEndArray();
+
+		gen.close();
+		return writer.toString();
+	}
+
+	/**
+	 * Extracts and aggregates all requested metrics from the given metric stores, and maps the result to a JSON string.
+	 *
+	 * @param stores available metrics
+	 * @param requestedMetrics ids of requested metrics
+	 * @param requestedAggregationsFactories requested aggregations
+	 * @return JSON string containing the requested metrics
+	 * @throws IOException
+	 */
+	private String getAggregatedMetricValues(Collection<? extends MetricStore.ComponentMetricStore> stores, String[] requestedMetrics, List<DoubleAccumulator.DoubleAccumulatorFactory<?>> requestedAggregationsFactories) throws IOException {
+		StringWriter writer = new StringWriter();
+		JsonGenerator gen = JsonFactory.JACKSON_FACTORY.createGenerator(writer);
+
+		gen.writeStartArray();
+		for (String requestedMetric : requestedMetrics) {
+			final Collection<Double> values = new ArrayList<>();
+			try {
+				for (MetricStore.ComponentMetricStore store : stores) {
+					String stringValue = store.metrics.get(requestedMetric);
+					if (stringValue != null) {
+						values.add(Double.valueOf(stringValue));
+					}
+				}
+			} catch (NumberFormatException nfe) {
+				// metric is not numeric so we can't perform aggregations => ignore it
+				continue;
```

Comment:
Not sure when this can happen, but shouldn't we log sth when we have this exception?
As in the `AggregatingSubtasksMetricsHandler.getStores()` in the `catch` part.

Reply:
this can happen for non-numeric metrics, for example the path to the last externalized checkpoint. Ill add a warn message.